packages/xforms-engine/test/parse/xpath/semantic-analysis.test.ts

```diff
@@ -109,4 +111,49 @@ describe('Semantic analysis', () => {
 			expect(actual).toEqual(expected);
 		});
 	});
+
+	interface ExpressionPredicateCase {
+		readonly expression: string;
+		readonly expected: boolean;
+	}
+
+	describe('Constant expressions', () => {
+		it.each<ExpressionPredicateCase>([
```

Comment:
following cases fail:

```
{ expression: 'not(false())', expected: true },
{ expression: 'number("1.2")', expected: true },
{ expression: 'boolean(1)', expected: true },
```

Similarly `isConstantTruthyExpression` would fail as well when a boolean returning function is used with constant arguments.

To me this is low priority and can be handled in a followup PR, I don't think people would be doing this kind of stuff in their Form definitions. Just documenting the limitations of these two functions would be sufficient.

Reply:
Yeah. The intent isn't really to be exhaustive, mainly to bring the existing usage (which is even more incomplete, and much more naive) in line with the overlapping functionality this change needed.

I'd also say that this feels like functionality we'll want to push down the stack. The usage in `createComputedExpression` is an optimization that would be beneficial to any user of `@getodk/xpath`. There are already quite a few internal optimizations in that package where certain statically known things are evaluated at "compile" time (in the sense that XPath expressions are parsed into evaluable representations).

For the non-optimization `isConstantTruthyExpression` usage, I think some of the cases introduced in this PR are already a bit of a stretch from how real world forms will be defined (typical usage being `readonly="true()"`). While it's definitely more "correct" to handle the cases you posted here, I don't know that they'd make any difference either way on real forms.